codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -43,32 +58,87 @@
+     * reside.
+     */
+    private final IntObjectMap<State> stateOnlyMap;
+    private final PriorityQueue<State> stateOnlyRemovalQueue;
+    private final Http2Connection connection;
     private final State connectionState;
     /**
      * The minimum number of bytes that we will attempt to allocate to a stream. This is to
      * help improve goodput on a per-stream basis.
      */
-    private int allocationQuantum = 1024;
+    private int allocationQuantum = DEFAULT_MIN_ALLOCATION_CHUNK;
+    private final int maxStateOnlySize;
 
     public WeightedFairQueueByteDistributor(Http2Connection connection) {
+        // FireFox currently uses 5 streams to establish QoS classes.
+        this(connection, 5);
+    }
+
+    public WeightedFairQueueByteDistributor(Http2Connection connection, int maxStateOnlySize) {
+        if (maxStateOnlySize < 0) {
+            throw new IllegalArgumentException("maxStateOnlySize: " + maxStateOnlySize + " (expected: >0)");
+        } else if (maxStateOnlySize == 0) {
+            stateOnlyMap = IntCollections.emptyMap();
+            stateOnlyRemovalQueue = EmptyPriorityQueue.instance();
+        } else {
+            stateOnlyMap = new IntObjectHashMap<State>(maxStateOnlySize);
+            // +2 because we may exceed the limit by 2 if a new dependency has no associated Http2Stream object. We need
+            // to create the State objects to put them into the dependency tree, which then impacts priority.
+            stateOnlyRemovalQueue = new DefaultPriorityQueue<State>(StateOnlyComparator.INSTANCE, maxStateOnlySize + 2);
+        }
+        this.maxStateOnlySize = maxStateOnlySize;
+
+        this.connection = connection;
         stateKey = connection.newKey();
-        Http2Stream connectionStream = connection.connectionStream();
+        final Http2Stream connectionStream = connection.connectionStream();
         connectionStream.setProperty(stateKey, connectionState = new State(connectionStream, 16));
 
         // Register for notification of new streams.
         connection.addListener(new Http2ConnectionAdapter() {
             @Override
             public void onStreamAdded(Http2Stream stream) {
-                stream.setProperty(stateKey, new State(stream));
+                State state = stateOnlyMap.remove(stream.id());
+                if (state == null) {
+                    state = new State(stream);
+                    // Only the stream which was just added will change parents. So we only need an array of size 1.
+                    List<ParentChangedEvent> events = new ArrayList<ParentChangedEvent>(1);
+                    connectionState.takeChild(state, false, events);
+                    notifyParentChanged(events);
+                } else {
+                    stateOnlyRemovalQueue.removeTyped(state);
+                    state.stream = stream;
+                }
+                switch (stream.state()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add a default case to this switch. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)